File: fail_over_project/app/schemas/media_request.py
from pydantic import BaseModel, Field

from app.models.enums import MediaRequestStatus, MediaRequestType


class MediaRequestOut(BaseModel):
    id: int
    user_id: int
    request_type: MediaRequestType
    media_id: int | None = None
    title: str
    description: str | None = None
    media_type_id: int | None = None
    status: MediaRequestStatus
    admin_note: str | None = None

    class Config:
        from_attributes = True


class MediaRequestCreate(BaseModel):
    request_type: MediaRequestType
    media_id: int | None = None
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    media_type_id: int | None = None


class MediaRequestAdminPatch(BaseModel):
    status: MediaRequestStatus
    admin_note: str | None = None

File: frontend/src/app/Services/media-request.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { firstValueFrom } from 'rxjs';

import { environment } from '../../environments/environment';

export type MediaRequestOut = {
  id: number;
  user_id: number;
  request_type: 'add' | 'update';
  media_id?: number | null;
  title: string;
  description?: string | null;
  media_type_id?: number | null;
  status: 'pending' | 'accepted' | 'resolving' | 'resolved' | 'rejected';
  admin_note?: string | null;
};

export type MediaRequestCreate = {
  request_type: 'add' | 'update';
  media_id?: number | null;
  title: string;
  description?: string | null;
  media_type_id?: number | null;
};

@Injectable({
  providedIn: 'root',
})
export class MediaRequest {
  constructor(private readonly http: HttpClient) {}

  async create(payload: MediaRequestCreate): Promise<MediaRequestOut> {
    return await firstValueFrom(
      this.http.post<MediaRequestOut>(`${environment.apiBaseUrl}/api/media-requests/`, payload),
    );
  }

  async listMine(): Promise<MediaRequestOut[]> {
    return await firstValueFrom(this.http.get<MediaRequestOut[]>(`${environment.apiBaseUrl}/api/media-requests/me`));
  }

  async listAll(): Promise<MediaRequestOut[]> {
    return await firstValueFrom(this.http.get<MediaRequestOut[]>(`${environment.apiBaseUrl}/api/media-requests/`));
  }

  async adminPatch(requestId: number, payload: { status: MediaRequestOut['status']; admin_note?: string | null }): Promise<MediaRequestOut> {
    return await firstValueFrom(
      this.http.patch<MediaRequestOut>(`${environment.apiBaseUrl}/api/media-requests/${requestId}`, payload),
    );
  }
}
